src/Features/Core/Portable/SolutionCrawler/IdleProcessor.cs

```diff
@@ -65,9 +66,23 @@ protected async Task WaitForIdleAsync()
                     return;
                 }
 
-                // TODO: will safestart/unwarp capture cancellation exception?
-                var timeLeft = BackOffTimeSpanInMS - diffInMS;
-                await Task.Delay(Math.Max(MinimumDelayInMS, timeLeft), this.CancellationToken).ConfigureAwait(continueOnCapturedContext: false);
+                var blockedOnCompletion = notifyBlockedOnCompletion.BlockedOnCompletion;
+                using (var combinedCancellationToken = CancellationToken.CombineWith(blockedOnCompletion))
+                {
+                    try
+                    {
+                        // TODO: will safestart/unwarp capture cancellation exception?
+                        var timeLeft = BackOffTimeSpanInMS - diffInMS;
+                        await Task.Delay(Math.Max(MinimumDelayInMS, timeLeft), combinedCancellationToken.Token).ConfigureAwait(false);
+                    }
+                    catch (OperationCanceledException) when (blockedOnCompletion.IsCancellationRequested && !CancellationToken.IsCancellationRequested)
```

Comment:
I'm really wishing the INotifyBlockedOnCompletion interface was instead an interface that gave you a Task.Delay-like method, and implemented the cancellation logic internally. More concretely:

```
interface IExpeditableDelaySource
{
    Task CreateExpeditableDelay(TimeSpan span, CancellationToken cancellationToken);
}
```

Then it means that the catching of the `BlockedOnCompletion` OperationCancelledException can be abstracted inside of it (it just is treated as a Delay that just completed sooner), which would greatly simplify this logic and also the logic at the other uses. It also means any new use doesn't have a chance of accidentally forgetting to dispose the combined token, won't incorrectly catch the OperationCanceled, etc.

Reply:
(and if you need to distinguish which was the cause of the delay, have it return a bool, which can be ignored.)

Reply:
💯 outstanding